auth/src/krb.cpp

```diff
@@ -0,0 +1,942 @@
+                                          const_cast<gss_OID>(gss_nt_service_name_krb),
+                                          &myName);
+            if (majorStatus != GSS_S_COMPLETE) {
+                const std::string msg = fmt::format(fmt::runtime(
+                                                    "Failed importing specified name: \"{}\"."),
+                                                    _specified_name);
+
+                krb_log_error(_r_error, msg.data(), majorStatus, minorStatus, _is_client);
+
+                THROW(KRB_ERROR_IMPORT_NAME, msg);
+            }
+        }
+
+        // Acquire the credentials if we do not have any yet
+        if (my_creds == GSS_C_NO_CREDENTIAL) {
+            majorStatus = gss_acquire_cred(
+                    &minorStatus,
+                    myName,
+                    0,
+                    GSS_C_NULL_OID_SET,
+                    _specified_name.empty() ? GSS_C_INITIATE : GSS_C_ACCEPT,
+                    &my_creds,
+                    NULL,
+                    NULL);
+        }
+        else {
+            // Already have credentials so all is well
+            majorStatus = GSS_S_COMPLETE;
+        }
+
+        if (majorStatus != GSS_S_COMPLETE) {
+            static const char* msg = "Failed acquiring credentials.";
+            krb_log_error(_r_error, msg, majorStatus, minorStatus, _is_client);
+            THROW(KRB_ERROR_ACQUIRING_CREDS, msg);
+        }
+
+        // set the credentials in the auth object
+        gss_release_name(&minorStatus, &myName);
+
+        majorStatus = gss_inquire_cred(&minorStatus, my_creds, &myName2, NULL, NULL, NULL);
+        if (majorStatus != GSS_S_COMPLETE) {
+            static const char* msg = "Failed inquiring creds for the name.";
+            krb_log_error(_r_error, msg, majorStatus, minorStatus, _is_client);
+            THROW(KRB_ERROR_ACQUIRING_CREDS, msg);
+        }
+
+        gss_OID doid2;
+        gss_buffer_desc client_name2;
+        majorStatus = gss_display_name(&minorStatus, myName2, &client_name2, &doid2);
+        if (majorStatus != GSS_S_COMPLETE) {
+            static const char* msg = "Failed during displaying name.";
+            krb_log_error(_r_error, msg, majorStatus, minorStatus, _is_client);
+            THROW(KRB_ERROR_DISPLAYING_NAME, msg);
+        }
+
+        std::string display_name;
+        if (client_name2.value && client_name2.length > 0) {
+            display_name = std::string{static_cast<char*>(client_name2.value)};
+        }
+
+        majorStatus = gss_release_name(&minorStatus, &myName2);
+        if (majorStatus != GSS_S_COMPLETE) {
+            static const char* msg = "Failed to release cred name.";
+            krb_log_error(_r_error, msg, majorStatus, minorStatus, _is_client);
+            THROW(KRB_ERROR_RELEASING_NAME, msg);
+        }
+
+        // TODO: should this be released?
+        gss_release_buffer(&minorStatus, &client_name2);
```

Comment:
Could capture in an issue and handle later. Especially if the original code did this as well.

Reply:
I added this TODO in the effort to remove `ASSERT_ERROR`/`ASSERT_PASS` recently because it seemed to me like we would always want to release the name, but was not the case before. It was only being done in the *success* code path and *failure* code paths would result in returning back with an error, never releasing the buffer and presumably leaking. See note about `gss_display_name`'s `output_name_buffer` here: https://www.gnu.org/software/gss/manual/html_node/Name-Manipulation.html#Name-Manipulation

I think an investigation issue would be warranted.

Reply:
I suppose this is also only releasing on success as we `THROW` on failure above.

Reply:
Releasing the buffer seems correct, but I agree we should investigate this before making any assumptions.

Reply:
Created https://github.com/irods/irods_auth_plugin_kerberos/issues/56 for this